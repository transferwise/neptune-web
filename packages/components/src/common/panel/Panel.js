import React, { useState, forwardRef } from 'react';
import Types from 'prop-types';
import classnames from 'classnames';
import { usePopper } from 'react-popper';

import { Position } from '..';
import './Panel.css';

const Panel = forwardRef(({ arrow, children, className, open, position, triggerRef }, ref) => {
  const [arrowElement, setArrowElement] = useState(null);
  const [popperElement, setPopperElement] = useState(null);

  const modifiers = [];

  if (arrow) {
    modifiers.push({ name: 'arrow', options: { padding: 5, element: arrowElement } });
    // This lets you displace a popper element from its reference element.
    modifiers.push({ name: 'offset', options: { offset: [0, 16] } });
  }

  const { styles, attributes } = usePopper(triggerRef.current, popperElement, {
    placement: position,
    modifiers,
  });

  return (
    // Popper recommends to use the popper element as a wrapper around an inner element that can have any CSS property transitioned for animations.
    <div
      ref={setPopperElement}
      style={{ ...styles.popper }}
      {...attributes.popper}
      className={classnames('np-panel')}
    >
      <div
        ref={ref}
        className={classnames('np-panel__content', { 'np-panel--open': open }, className)}
      >
        {children}
      </div>
      {arrow && <div className="np-panel__arrow" ref={setArrowElement} style={styles.arrow} />}
    </div>
  );
});

Panel.Position = {
  BOTTOM: Position.BOTTOM,
  LEFT: Position.LEFT,
  RIGHT: Position.RIGHT,
  TOP: Position.TOP,
};

Panel.defaultProps = {
  arrow: true,
  className: undefined,
  open: false,
  position: Panel.Position.TOP,
};

Panel.propTypes = {
  arrow: Types.bool,
  className: Types.string,
  children: Types.node.isRequired,
  open: Types.bool,
  position: Types.oneOf([
    Panel.Position.BOTTOM,
    Panel.Position.LEFT,
    Panel.Position.RIGHT,
    Panel.Position.TOP,
  ]),
  triggerRef: Types.shape({ current: Types.shape({}) }).isRequired,
};

export default Panel;
