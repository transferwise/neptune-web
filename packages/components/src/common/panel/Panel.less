@import (reference) '@transferwise/neptune-css/src/less/variables/_typography.less';
@import (reference) '@transferwise/neptune-css/dist/less/zindex.less';

.np-panel {
  z-index: @zindex-popover;
  &__content {
    opacity: 0;
    visibility: hidden;
    padding: var(--size-16) var(--size-24);
    background-color: var(--color-background-elevated);
    border-radius: var(--radius-small);
    box-shadow: 0px 0px 40px rgba(34, 48, 73, 0.2);
    transition: opacity 0.2s cubic-bezier(0.6, 0.2, 0.1, 1),
      transform 0.2s cubic-bezier(0.6, 0.2, 0.1, 1), visibility 0s linear 0s;
    transform: translateY(-10px) scale(0.95);
  }
  &--open {
    transform: none;
    opacity: 1;
    visibility: visible;
  }

  &__arrow,
  &__arrow::before {
    position: absolute;
    width: 14px;
    height: 14px;
  }

  &__arrow::before {
    content: '';
    transform: rotate(45deg);
    background: var(--color-background-elevated);
  }

  &[data-popper-placement^='top'] > &__content &__arrow {
    bottom: -6px;
  }

  &[data-popper-placement^='right'] > &__content &__arrow {
    left: -6px;
  }

  &[data-popper-placement^='bottom'] > &__content &__arrow {
    top: -6px;
  }

  &[data-popper-placement^='left'] > &__content &__arrow {
    right: -6px;
  }
}
