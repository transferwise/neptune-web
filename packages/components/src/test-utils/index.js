import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 */
function customRender(ui, { locale = DEFAULT_LOCALE, messages = en, ...renderOptions } = {}) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return <Provider i18n={{ locale, messages }}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render };
