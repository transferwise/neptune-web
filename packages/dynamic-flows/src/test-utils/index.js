import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';
import { act } from 'react-dom/test-utils';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount, getMockFetchPromise, wait };
