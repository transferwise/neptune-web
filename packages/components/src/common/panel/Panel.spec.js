import React from 'react';
import { render } from '../../test-utils';
import Panel from './Panel';

describe('Panel', () => {
  const props = {
    arrow: true,
    children: <div>children</div>,
    open: false,
    position: Panel.Position.TOP,
    anchorRef: {},
  };

  it('renders', () => {
    const { container } = render(<Panel {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('renders arrow', () => {
    const { rerender } = render(<Panel {...props} arrow={false} />);

    expect(getArrow()).not.toBeInTheDocument();

    rerender(<Panel {...props} />);

    expect(getArrow()).toBeInTheDocument();
  });

  it('opens/closes Panel based on open', async () => {
    const { rerender } = render(<Panel {...props} open={false} />);

    expect(getPanel()).not.toBeInTheDocument();

    rerender(<Panel {...props} open />);

    expect(getPanel()).toBeInTheDocument();
  });

  const getPanel = () => document.querySelector('.np-panel--open');
  const getArrow = () => document.querySelector('.np-panel__arrow');
});
