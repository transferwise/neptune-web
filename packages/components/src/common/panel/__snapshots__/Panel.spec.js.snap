// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel renders 1`] = `
<div>
  <div
    class="np-panel"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="np-panel__content"
    >
      <div>
        children
      </div>
    </div>
    <div
      class="np-panel__arrow"
    />
  </div>
</div>
`;
